node/node.go

```diff
@@ -587,6 +636,7 @@ func (n *Node) delFeed(pk cipher.PubKey) (ok bool) {
 
 	if _, ok = n.feeds[pk]; ok {
 		delete(n.feeds, pk)
+		go n.discovery.ForEachConn(n.updateServiceDiscovery)
```

Comment:
Every of this goroutines will wait until the `delFeed` finish. Because, the `Feeds` method locks the `fmx`. Thus, for 10000 connections it launch 10000 waiting goroutines, everyone of which calls `s.Feeds()` that allocates slice. Every time.

Try something like:
```go
// after the defer


if _, ok = n.feeds[pk]; !ok {
    return
}
delete(n.feeds, pk)

feeds := make([]cipher.PubKey, 0, len()n.feeds)
services := make([]*factory.Service, 0, len(feeds))

for pk := range n.feeds {
    feeds = append(feeds, pk)
    services = append(services, &factory.Service{Key: pk})
}
n.discovery.ForEachConn(func(conn *factory.Conn) {
    n.updateServiceDiscoverySomeOtherNameActually(feeds, services)
})
```
And the `updateServiceDiscovery` (it used as callback somewhere) will calls this `updateServiceDiscoverySomeOtherNameActually` method with the same arguments.
E.g. this `updateServiceDiscoverySomeOtherNameActually` don't lock the fmx.

And also, do you know that a node loads list of feeds from database during start? Do you handle it?

Reply:
I think may be you could have a chance to improve your exported `Feeds`, like using copy on write algorithm?
Then I can reuse your code and every user can use it not so carefully.

Reply:
> And also, do you know that a node loads list of feeds from database during start? Do you handle it?

Yes

Reply:
The cow slice is not safe. It requires "DON'T EDIT" note. But I will add it. Ha-ha. Thanks.